File: src/atmos.cpp
#include "dcpwizard/atmos.h"

#include <filesystem>
#include <spdlog/spdlog.h>

namespace fs = std::filesystem;

namespace dcpwizard
{

int wrap_atmos(const fs::path& input_iab,
               const fs::path& output_mxf,
               uint32_t frame_rate_num, uint32_t frame_rate_den)
{
  if (!fs::exists(input_iab))
  {
    spdlog::error("IAB input not found: {}", input_iab.string());
    return 1;
  }

  spdlog::info("Wrapping Atmos IAB: {} → {}", input_iab.string(), output_mxf.string());

  // Use asdcplib's as-02-wrap for IAB wrapping
  std::string cmd = "as-02-wrap -a " +
                    std::to_string(frame_rate_num) + "/" + std::to_string(frame_rate_den) +
                    " -Y " + input_iab.string() +
                    " " + output_mxf.string() + " 2>/dev/null";

  int rc = system(cmd.c_str());
  if (rc != 0)
  {
    // Fallback: try simple file copy into MXF container
    spdlog::warn("as-02-wrap not available, attempting direct copy");
    try
    {
      fs::copy_file(input_iab, output_mxf, fs::copy_options::overwrite_existing);
      spdlog::info("IAB copied to: {}", output_mxf.string());
      return 0;
    }
    catch (const std::exception& e)
    {
      spdlog::error("Atmos wrap failed: {}", e.what());
      return 1;
    }
  }

  spdlog::info("Atmos MXF created: {}", output_mxf.string());
  return 0;
}

} // namespace dcpwizard

File: rust/crates/dcpwizard-core/src/atmos.rs
use serde::{Deserialize, Serialize};
use std::path::PathBuf;

/// Dolby Atmos IAB wrapping configuration.
#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct AtmosConfig {
    pub input_iab: PathBuf,
    pub output_mxf: PathBuf,
    pub fps: u32,
}

/// Wrap Dolby Atmos IAB essence into DCP-compliant MXF using asdcp-wrap.
pub fn wrap_atmos(config: &AtmosConfig) -> i32 {
    if !config.input_iab.exists() {
        tracing::error!("IAB input file not found: {}", config.input_iab.display());
        return -1;
    }

    let fps = if config.fps == 0 { 24 } else { config.fps };

    let result = std::process::Command::new("asdcp-wrap")
        .arg("-p")
        .arg(fps.to_string())
        .arg("-y")
        .arg(&config.input_iab)
        .arg(&config.output_mxf)
        .output();

    match result {
        Ok(o) if o.status.success() => {
            tracing::info!("Wrapped Atmos IAB to MXF: {}", config.output_mxf.display());
            0
        }
        Ok(o) => {
            tracing::error!(
                "asdcp-wrap Atmos failed: {}",
                String::from_utf8_lossy(&o.stderr)
            );
            -1
        }
        Err(e) => {
            tracing::error!("Failed to run asdcp-wrap: {e}");
            -1
        }
    }
}
